round-based/src/simulation/mod.rs

```diff
@@ -1,49 +1,496 @@
+//!     n,
+//!     |i, party| protocol_of_random_generation(party, i, n),
+//! )
+//! .await
+//! // unwrap `Result`s
+//! .expect_ok()
+//! // check that all parties produced the same response
+//! .expect_eq();
+//!
+//! println!("Output randomness: {}", hex::encode(output));
+//! # }  
 //! ```
 
-mod sim_async;
-#[cfg(feature = "state-machine")]
-mod sim_sync;
+use alloc::{boxed::Box, collections::VecDeque, vec::Vec};
+use core::future::Future;
+
+use crate::{state_machine::ProceedResult, Incoming, MessageDestination, MessageType, Outgoing};
+
+#[cfg(feature = "sim-async")]
+pub mod async_env;
+
+/// Result of the simulation
+pub struct SimResult<T>(pub Vec<T>);
+
+impl<T, E> SimResult<Result<T, E>>
+where
+    E: core::fmt::Debug,
+{
+    /// Unwraps `Result<T, E>` produced by each party
+    ///
+    /// Panics if at least one of the parties returned `Err(_)`. In this case,
+    /// a verbose error message will shown specifying which of the parties returned
+    /// an error.
+    pub fn expect_ok(self) -> SimResult<T> {
+        let mut oks = Vec::with_capacity(self.0.len());
+        let mut errs = Vec::with_capacity(self.0.len());
+
+        for (res, i) in self.0.into_iter().zip(0u16..) {
+            match res {
+                Ok(res) => oks.push(res),
+                Err(res) => errs.push((i, res)),
+            }
+        }
+
+        if !errs.is_empty() {
+            let mut msg = alloc::format!(
+                "Simulation output didn't match expectations.\n\
+                Expected: all parties succeed\n\
+                Actual  : {success} parties succeeded, {failed} parties returned an error\n\
+                Failures:\n",
+                success = oks.len(),
+                failed = errs.len(),
+            );
+
+            for (i, err) in errs {
+                msg += &alloc::format!("- Party {i}: {err:?}\n");
+            }
+
+            panic!("{msg}");
+        }
+
+        SimResult(oks)
+    }
+}
+
+impl<T> SimResult<T>
+where
+    T: PartialEq + core::fmt::Debug,
+{
+    /// Checks that outputs of all parties are equally the same
+    ///
+    /// Returns the output on success (all the outputs are checked to be the same), otherwise
+    /// panics with a verbose error message.
+    ///
+    /// Panics if simulation contained zero parties.
+    pub fn expect_eq(mut self) -> T {
+        let Some(first) = self.0.first() else {
+            panic!("simulation contained zero parties");
+        };
+
+        if !self.0[1..].iter().all(|i| i == first) {
+            let mut msg = alloc::string::String::from(
+                "Simulation output didn't match expectations.\n\
+                Expected: all parties return the same output\n\
+                Actual  : some of the parties returned a different output\n\
+                Outputs :\n",
+            );
+
+            for (i, res) in self.0.iter().enumerate() {
+                msg += &alloc::format!("- Party {i}: {res:?}");
+            }
```

Comment:
Might be helpful to cluster results by equality, so it would be like `- Parties 1, 2: {res1:?} \n - Parties 3, 4: {res2:?}`